rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+---
+feature: pipe-operator
+start-date: 2023-05-23
+author: @piegamesde
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Introduce a new "pipe" operator, `|>`, to the Nix language, defined as `f a` = `a |> f`.
+Additionally, elevate `lib.pipe` to a built-in function.
+
+As a reminder, `pipe a [ f g h ]` is defined as `h (g (f a))`.
+
+# Motivation
+[motivation]: #motivation
+
+Creating advanced data processing like transforming a list is a thing commonly done in nixpkgs.
+Yet the language has no support for function concatentation/composition,
+which results in such constructs looking unwieldy and difficult to format well.
+`lib.pipe` may be the most powerful library function with that regard,
+but it is unknown and overlooked by many because it is not easily discoverable:
+Despite its great usefulness, it is currently used in less than 30 files in Nixpkgs
+(`rg '[\. ]pipe .* \['`).
+Additionally, it is not accessible to Nix code outside of nixpkgs,
+and due to Nix's lazy evaluation debugging type errors is really difficult.
+
+Let's have a look at an arbitrarily chosen snippet of Nixpkgs code:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (lib.concatStringsSep "\n" (lib.mapAttrsToList (key: value: ''
+  // ${value.reason}
+  pref("${key}", ${builtins.toJSON value.value});
+'') defaultPrefs));
+```
+
+It is arguably pretty hard to read and reason about. Even when applying some more whitespace-generous formatting:
+
+```nix
+defaultPrefsFile = pkgs.writeText "nixos-default-prefs.js" (
+  lib.concatStringsSep "\n" (
+    lib.mapAttrsToList
+    (
+      key: value: ''
+        // ${value.reason}
+        pref("${key}", ${builtins.toJSON value.value});
+      ''
+    )
+    defaultPrefs
+  )
+);
+```
+
+One can observe the following issues:
+
+- If you want to follow the data flow, you must read it from bottom to top,
+  from the inside to the outside (the input here is `defaultPrefs`).
+- Adding a function call to the output would require wrapping the entire
+  expression in parentheses and increasing its indentation.
+
+Compare this to the equivalent call with `lib.pipe`:
+
+```nix
+defaultPrefsFile = pipe defaultPrefs [
+  (lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  ))
+  (lib.concatStringsSep "\n")
+  (pkgs.writeText "nixos-default-prefs.js")
+];
+```
+
+The code now clearly reads from top to bottom in the order the data is processed,
```

Comment:
While I agree that the code reads clearly, I think the "top to bottom" part is somewhat of a detriment, actually. One of the first things that a beginner learns about Nixlang is how the order of statements DOES NOT matter. Of course, you *should* write code from top to bottom, but it is not always possible or feasible, and Nixlang itself doesn't care about the ordering.

In practice, this means that a lot of the time while reading Nixlang code, you have to make mental notes about the stuff you haven't yet seen the definition or usage of, or jump around the code in the `goto`-like fashion. `pipe` being one of the exceptions where you MUST read from top to bottom is actually unintuitive here.

Reply:
> One of the first things that a beginner learns about Nixlang is how the order of statements DOES NOT matter.

Of course the order of statements does matter, what you mean here is that due to Nix being a pure lazy functional language the *execution* order of statements is not strictly dictated by their appearance order in code. Therefore it is even more important to focus on the ordering relation that actually does matter in practice, and for me that is data flow.

> In practice, this means that a lot of the time while reading Nixlang code, you have to make mental notes about the stuff you haven't yet seen the definition or usage of, or jump around the code in the goto-like fashion. pipe being one of the exceptions where you MUST read from top to bottom is actually unintuitive here.

I find this statement weird, because it could equally well also apply to any function application itself: "function application being one of the exceptions where you MUST read from the inside to the outside is actually unintuitive here"

If you care about the direction, there is always the possibility to introduce `<|` instead, which is equivalent to basic function application.